runtime-light/stdlib/rpc/rpc-api.h

```diff
@@ -6,93 +6,208 @@
+  return true;
+}
 
-// === Rpc Fetch ==================================================================================
+inline void f$store_raw_vector_double(const array<double>& vector) noexcept {
+  const std::string_view vector_view{reinterpret_cast<const char*>(vector.get_const_vector_pointer()), sizeof(double) * vector.count()};
+  RpcInstanceState::get().rpc_buffer.store_bytes(vector_view);
+}
 
-int64_t f$fetch_int() noexcept;
+inline int64_t f$fetch_int() noexcept {
+  return static_cast<int64_t>(RpcInstanceState::get().rpc_buffer.fetch_trivial<int32_t>().value_or(0));
+}
 
-int64_t f$fetch_long() noexcept;
+inline int64_t f$fetch_long() noexcept {
+  return RpcInstanceState::get().rpc_buffer.fetch_trivial<int64_t>().value_or(0);
+}
 
-double f$fetch_double() noexcept;
+inline double f$fetch_double() noexcept {
+  return RpcInstanceState::get().rpc_buffer.fetch_trivial<double>().value_or(0.0);
+}
 
-double f$fetch_float() noexcept;
+inline double f$fetch_float() noexcept {
+  return static_cast<double>(RpcInstanceState::get().rpc_buffer.fetch_trivial<float>().value_or(0));
+}
 
-string f$fetch_string() noexcept;
+inline string f$fetch_string() noexcept {
+  tl::string str{};
+  str.fetch(RpcInstanceState::get().rpc_buffer);
+  return {str.value.data(), static_cast<string::size_type>(str.value.size())};
+}
 
-// === Rpc Query ==================================================================================
+inline void f$fetch_raw_vector_double(array<double>& vector, int64_t num_elems) noexcept {
+  auto& rpc_buf{RpcInstanceState::get().rpc_buffer};
+  const auto len_bytes{sizeof(double) * num_elems};
+  if (rpc_buf.remaining() < len_bytes) [[unlikely]] {
+    return; // TODO: error handling
+  }
+  vector.memcpy_vector(num_elems, std::next(rpc_buf.data(), rpc_buf.pos()));
+  rpc_buf.adjust(len_bytes);
+}
+
+inline void f$rpc_clean() noexcept {
+  RpcInstanceState::get().rpc_buffer.clean();
+}
+
+template<typename T>
+bool f$rpc_parse(T /*unused*/) {
+  php_critical_error("call to unsupported function");
+}
+
+// === client =====================================================================================
+
+inline int64_t f$rpc_tl_pending_queries_count() noexcept {
+  return RpcInstanceState::get().response_waiter_forks.size();
+}
 
-kphp::coro::task<array<int64_t>> f$rpc_send_requests(string actor, array<mixed> tl_objects, Optional<double> timeout, bool ignore_answer,
-                                                     class_instance<C$KphpRpcRequestsExtraInfo> requests_extra_info, bool need_responses_extra_info) noexcept;
+// === client untyped =============================================================================
+
+inline kphp::coro::task<array<int64_t>> f$rpc_send_requests(string actor, array<mixed> tl_objects, Optional<double> timeout, bool ignore_answer,
+                                                            class_instance<C$KphpRpcRequestsExtraInfo> requests_extra_info,
```

Comment:
I think it should be defined in .cpp file

Reply:
I don't see a reason for that